src/Features/Core/Portable/ConvertIfToSwitch/AbstractConvertIfToSwitchCodeRefactoringProvider.cs

```diff
@@ -1,256 +1,124 @@
 
+#nullable enable
+
 using System;
-using System.Collections.Generic;
+using System.Collections.Immutable;
 using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
 using Microsoft.CodeAnalysis.CodeActions;
 using Microsoft.CodeAnalysis.CodeRefactorings;
-using Microsoft.CodeAnalysis.Editing;
-using Microsoft.CodeAnalysis.Formatting;
 using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Operations;
 using Microsoft.CodeAnalysis.Shared.Extensions;
 
 namespace Microsoft.CodeAnalysis.ConvertIfToSwitch
 {
-    internal abstract class AbstractConvertIfToSwitchCodeRefactoringProvider : CodeRefactoringProvider
+    internal abstract partial class AbstractConvertIfToSwitchCodeRefactoringProvider<
+        TIfStatementSyntax, TExpressionSyntax, TIsExpressionSyntax, TPatternSyntax> : CodeRefactoringProvider
     {
         public sealed override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
         {
             var document = context.Document;
+            var cancellationToken = context.CancellationToken;
+
             if (document.Project.Solution.Workspace.Kind == WorkspaceKind.MiscellaneousFiles)
             {
                 return;
             }
 
-            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
-            var semanticModel = await document.GetSemanticModelAsync().ConfigureAwait(false);
-
-            await CreateAnalyzer(syntaxFacts, semanticModel)
-                .ComputeRefactoringsAsync(context).ConfigureAwait(false);
-        }
-
-        protected interface IPattern<TSwitchLabelSyntax>
-            where TSwitchLabelSyntax : SyntaxNode
-        {
-            TSwitchLabelSyntax CreateSwitchLabel();
-        }
-
-        protected interface IAnalyzer
-        {
-            Task ComputeRefactoringsAsync(CodeRefactoringContext context);
-        }
-
-        protected abstract IAnalyzer CreateAnalyzer(ISyntaxFactsService syntaxFacts, SemanticModel semanticModel);
-
-        protected abstract class Analyzer<TStatementSyntax, TIfStatementSyntax, TExpressionSyntax, TSwitchLabelSyntax> : IAnalyzer
-            where TExpressionSyntax : SyntaxNode
-            where TIfStatementSyntax : SyntaxNode
-            where TSwitchLabelSyntax : SyntaxNode
-        {
-            protected readonly ISyntaxFactsService _syntaxFacts;
-            protected readonly SemanticModel _semanticModel;
-            private int _numberOfSubsequentIfStatementsToRemove = -1;
-            private TExpressionSyntax _switchExpression;
-            private Optional<TStatementSyntax> _switchDefaultBodyOpt;
-
-            public Analyzer(ISyntaxFactsService syntaxFacts, SemanticModel semanticModel)
+            var ifStatement = await context.TryGetRelevantNodeAsync<TIfStatementSyntax>().ConfigureAwait(false);
+            if (ifStatement == null || ifStatement.ContainsDiagnostics)
```

Comment:
`ContainsDiagnostics` isn't great.  even an errant warning deep down in the tree will trigger this.